vmdb/app/models/orchestration_stack.rb

```diff
@@ -40,25 +40,39 @@ def total_cloud_networks
     cloud_networks.size
   end
 
-  # @param options [Hash] what to update for the stack. Option keys and values are:
-  #   :template (String, URI, S3::S3Object, Object) - A new stack template.
-  #     This may be provided in a number of formats including:
-  #       a String, containing the template in CFN or HOT format.
-  #       a URL String pointing to the document in S3.
-  #       a URI object pointing to the document in S3.
-  #       an S3::S3Object which contains the template.
-  #       an Object which responds to #to_json and returns the template.
-  #   :parameters (Hash) - A hash that specifies the input parameters of the new stack.
+  def self.create_stack(orchestration_manager, stack_name, template, options = {})
+    klass = "OrchestrationStack#{orchestration_manager.class.name.match(/Ems(.*)/)[1]}".constantize
+    ems_ref = klass.raw_create_stack(orchestration_manager, stack_name, template, options)
+
+    klass.create(:name                   => stack_name,
+                 :ems_ref                => ems_ref,
+                 :status                 => 'CREATE_IN_PROGRESS',
+                 :ext_management_system  => orchestration_manager,
+                 :orchestration_template => template)
+  end
+
+  def self.raw_create_stack(_orchestration_manager, _stack_name, _template, _options = {})
+    raise NotImplementedError, "raw_create_stack must be implemented in a subclass"
+  end
```

Comment:
@Fryguy I'm definitely wondering the same thing... but that seems like a problem for Future Us (suckers!).

If we need to introduce duplication while splitting the providers, we'll do that trivially. Re-unifying code that's been deliberately duplicated is much more painful / less likely to happen just because it can.

I think the consistency is going to be pretty.. consistent, as long as they're all trying to put the data in a same-looking DB.